### PR TITLE
Do not throw exception when flat_object field is explicitly null

### PR DESCRIPTION
**Pull request by msfroh**
Backport of https://github.com/opensearch-project/OpenSearch/pull/15375/

### Description

It is valid for a flat_object field to have an explicit value of null. (It's functionally the same as not specifying the field at all.) Prior to this fix, though, we would erroneously advance the context parser to the next token, violating the contract with DocumentParser (which says that a call to parseCreateField with a null value should complete with the parser still pointing at the null value -- it is DocumentParser's responsibility to advance).

Signed-off-by: Michael Froh <froh@amazon.com>

* Fix unit test

Signed-off-by: Michael Froh <froh@amazon.com>

* Add changelog entry

Signed-off-by: Michael Froh <froh@amazon.com>

---------

Signed-off-by: Michael Froh <froh@amazon.com>
(cherry picked from commit 46a269ef21e88e3cb1398474e4bf82af4c3b3b7f)

### Related Issues
N/A

### Check List
- [x] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by jainankitk — APPROVED** [review]

**Review by reta — APPROVED** [review]

File: rest-api-spec/src/main/resources/rest-api-spec/test/index/100_partial_flat_object.yml [hunk 1]
```diff
@@ -88,7 +88,16 @@ setup:
             } ]
```
Comment (reta) [inline rest-api-spec/src/main/resources/rest-api-spec/test/index/100_partial_flat_object.yml:88]: We have to change `skip` to `2.16.99` here, right?

Comment (msfroh) [reply rest-api-spec/src/main/resources/rest-api-spec/test/index/100_partial_flat_object.yml:88]: I was going to make that change on main let auto-backport do its thing.

Comment (msfroh) [reply rest-api-spec/src/main/resources/rest-api-spec/test/index/100_partial_flat_object.yml:88]: I'll open that PR right now

Comment (msfroh) [reply rest-api-spec/src/main/resources/rest-api-spec/test/index/100_partial_flat_object.yml:88]: There we go: https://github.com/opensearch-project/OpenSearch/pull/15448

("Right now" became an hour as I got distracted.)

**Review by reta — APPROVED** [review]

**Pull request merged**